samples/drivers/uart/async_api/src/main.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2025 Embeint Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/net_buf.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/random/random.h>
+
+/* change this to any other UART peripheral if desired */
+#define UART_DEVICE_NODE DT_CHOSEN(zephyr_shell_uart)
+
+/* Maximum number of packets to generate per iteration */
+#define LOOP_ITER_MAX_TX 4
+/* Maximum size of our TX packets */
+#define MAX_TX_LEN 32
+#define RX_CHUNK_LEN 32
+
+/* Buffer pool for our TX payloads */
+NET_BUF_POOL_DEFINE(tx_pool, LOOP_ITER_MAX_TX, MAX_TX_LEN, 0, NULL);
+
+struct k_fifo tx_queue;
+struct net_buf *tx_pending_buffer;
+uint8_t async_rx_buffer[2][RX_CHUNK_LEN];
+volatile uint8_t async_rx_buffer_idx;
+
+static const struct device *const uart_dev = DEVICE_DT_GET(UART_DEVICE_NODE);
+
+LOG_MODULE_REGISTER(sample, LOG_LEVEL_INF);
+
+static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
+{
+	struct net_buf *buf;
+	int rc;
+
+	LOG_DBG("EVENT: %d", evt->type);
+
+	switch (evt->type) {
+	case UART_TX_DONE:
+		LOG_DBG("TX complete %p", tx_pending_buffer);
+
+		/* Free TX buffer */
+		net_buf_unref(tx_pending_buffer);
+		tx_pending_buffer = NULL;
+
+		/* Handle any queued buffers */
+		buf = k_fifo_get(&tx_queue, K_NO_WAIT);
+		if (buf != NULL) {
+			rc = uart_tx(dev, buf->data, buf->len, 0);
+			if (rc != 0) {
+				LOG_ERR("TX from ISR failed (%d)", rc);
+				net_buf_unref(buf);
+			} else {
+				tx_pending_buffer = buf;
+			}
+		}
+		break;
+	case UART_RX_BUF_REQUEST:
+		/* Return the next buffer index */
+		LOG_DBG("Providing buffer index %d", async_rx_buffer_idx);
+		rc = uart_rx_buf_rsp(dev, async_rx_buffer[async_rx_buffer_idx],
```

Comment:
it will assert with current implementation

Reply:
> It won't, because the native driver never issues UART_RX_BUF_REQUEST, since it has no need for multiple buffers.

But this is a generic sample for all UART drivers supporting the ASYNC API, or?

Reply:
My comment was in reference to what I thought @nordic-krch was referring to, which was that `uart_rx_buf_rsp` would return `-ENOTSUP` for the native driver, and thus trigger the assertion. If that is not the case I would need clarification on the "will assert" comment.